src/explosion.cpp

```diff
@@ -435,93 +435,116 @@ std::vector<tripoint> game::shrapnel( const tripoint &src, int power,
+                             critter->disp_name().c_str(),
+                             frag.proj.speed, rl_dist( src, target ) );
+                    add_msg( m_debug, "Shrapnel dealt %d damage", frag.dealt_dam.total_damage() );
+                    if( critter->is_dead_state() ) {
+                        break;
                     }
                 }
-                if( m.impassable( target ) ) {
-                    if( optional_vpart_position vp = m.veh_at( target ) ) {
-                        vp->vehicle().damage( vp->part_index(), damage );
+                int total_hits = damaging_hits + non_damaging_hits;
+                if( total_hits > 0 && g->u.sees( *critter ) ) {
+                    // Building a phrase to summarize the fragment effects.
+                    // Target, Number of impacts, total amount of damage, proportion of deflected fragments.
+                    std::map<int, std::string> impact_count_descriptions = {
+                        { 1, _( "a" ) }, { 2, _( "several" ) }, { 5, _( "many" ) },
+                        { 20, _( "a large number of" ) }, { 100, _( "a huge number of" ) },
+                        { std::numeric_limits<int>::max(), _( "an immense number of" ) }
+                    };
+                    std::string impact_count = std::find_if(
+                                                   impact_count_descriptions.begin(), impact_count_descriptions.end(),
+                    [ total_hits ]( std::pair<int, std::string> desc ) {
+                        return desc.first >= total_hits;
+                    } )->second;
+                    std::string damage_description = ( damage_taken > 0 ) ?
+                                                     string_format( _( "dealing %d damage" ), damage_taken ) :
+                                                     _( "but they deal no damage" );
+                    if( critter->is_player() ) {
+                        add_msg( ngettext( "You are hit by %s bomb fragment, %s.",
+                                           "You are hit by %s bomb fragments, %s.", total_hits ),
+                                 impact_count.c_str(), damage_description );
+                    } else if( critter->is_npc() ) {
+                        critter->add_msg_if_npc(
+                            ngettext( "<npcname> is hit by %s bomb fragment, %s.",
+                                      "<npcname> is hit by %s bomb fragments, %s.",
+                                      total_hits ),
+                            impact_count, damage_description );
                     } else {
-                        m.bash( target, damage / 10, true );
+                        add_msg( ngettext( "The %s is hit by %s bomb fragment, %s.",
+                                           "The %s is hit by %s bomb fragments, %s.", total_hits ),
+                                 critter->disp_name().c_str(), impact_count, damage_description );
```

Comment:
I just noticed this creates lines like "The the zombie is hit..." and "The the crawling zombie....". I think you wanted `critter->get_name()` instead?